### DISCUSSION TITLE
Using an external TheHive/Cortex installation

### DISCUSSION THREAD

**Discussion by achillean00** [Security-Onion-Solutions/securityonion#2576] [Unsupported Versions]
Hi,

I was wondering if it's possible to configure SO to use an external TheHive install rather than the bundled one? Nosing around the config files I noticed that SOCtopus.conf had definitions for URLs and API keys so it kind of feels like it might be possible?

Cheers, Ian

**Comment by TOoSmOotH** [comment]
In theory it might be possible. You would need to replace the URLs and API key in the pillars and see what happens. If you give it a try let me know and I can document it.

**Reply by achillean00** [reply]
Okay, after some fiddling and debugging I *nearly* have this working, and I think it will work fine for anyone running TheHive v3, more details on the issue I found later.

To point SO at an external TheHive instance I did the following:
Copy `SOCtopus.conf` from `/opt/so/saltstack/default/salt/soctopus/files` to `/opt/so/saltstack/local/salt/soctopus/files`
Edit `SOCtopus.conf` do the following:
Add the line **{%- set HIVEBASE = salt['pillar.get']('global:hive_base', '') %}** at the top to pull in the Hive base
Adjust the **hive_url** and **cortex_url** definitions to use the HIVEBASE variable. I needed to adjust the URL too as we don't use /thehive

Copy `soc.json` from `/opt/so/saltstack/default/salt/soc/files/soc` to `/opt/so/saltstack/local/salt/soc/files/soc`
Edit `soc.json` and do the following:
Add the line **{%- set THEHIVEBASE = salt['pillar.get']('global:hive_base', '') %}** at the top
Adjust the line  "hostUrl": "https://{{ THEHIVEURL }}", under thehive as needed. Again, I removed the /thehive

Edit `/opt/so/saltstack/local/pillar/global.sls` and add a variable **hive_base:** pointing at the hostname of the TheHive instance
Also adjust the hivekey: and cortexorguserkey: variables to include API keys for the external instance.

Update the configuration files by running `salt <manager_name>_manager state.highstate`

I imagine the caveat with doing this is that by overriding a couple of the configuration templates locally it means we'd have to do more checks after an upgrade as we wouldn't pick up changes to the default files.

After doing this I found that when clicking the escalate button in the SO dashboard I got a 400 error. I turned on some debugging on my TheHive instance and found that when SO submitted a new case it was generating a 404 error. Digging into this, and testing case generation using curl, I found that when SO submits a case it sends something like:

```
x.x.x.x -  [12/Jan/2021:10:18:49 +0000] "<hive_host>/api/case" "POST /api/case HTTP/1.1" 404 238 "" "Go-http-client/1.1" "{\"title\":\"ET CINS Active Threat Intelligence Poor R
eputation IP group 73\",\"description\":\"{\\\"count\\\":18,\\\"rule.name\\\":\\\"ET CINS Active Threat Intelligence Poor Reputation IP group 73\\\",\\\"event.module\\\":\\\"suricata\\\",\\\"eve
nt.severity_label\\\":\\\"medium\\\"}\",\"severity\":2,\"tags\":[\"SecurityOnion\"],\"tlp\":2,\"flag\":false,\"template\":\"\"}"
```
and TheHive is generating the following in the application.log
```
2021-01-12 10:18:49,577 [WARN] from org.thp.scalligraph.ErrorHandler in application-akka.actor.default-dispatcher-4091 [00000393|] POST /api/case returned 404
org.thp.scalligraph.NotFoundError: CaseTemplate not found
```

I tested this using curl against the SO instance and ours by doing:
```
curl -k -XPOST -H 'Authorization: Bearer <apitoken>' -H 'Content-Type: application/json' <hive_host>/api/case -d '{
  "title": "My first case",
  "description": "This case has been created by my custom script",
  "severity": 3
}'
```
and then
```
curl -k -XPOST -H 'Authorization: Bearer <apitoken>' -H 'Content-Type: application/json' <hive_host>/api/case -d '{
  "title": "My first case",
  "description": "This case has been created by my custom script",
  "severity": 3,
  "template": ""
}'
```
Against the SO instance both of these work fine, against our instance the second test throws the same 404 error.

We're running TheHive 4.0.3-1 so I suspect they've tightened up how cases are created by the API and now won't allow you to specify "template" as an empty declaration. I couldn't actually work out where you were adding template: to the JSON data as it didn't seem to be in the SOCtopus files. :) Might be something worth watching out for if you move to TheHive4 in the future.

**Reply by weslambert** [reply]
Thanks for trying this out.

We've held off on moving to TheHive4 due to the fact that Cortex and TheHive would not be using the same backend database.   We may consider adjusting this to allow compatibility for folks running an external instance of TheHive4, but for now, it's not really a priority. 

Out of curiosity, with TH4, will it generate an error if you submit a template that does not exist ( does it have to match an existing one, or will it ignore it)?

**Reply by achillean00** [reply]
TH4 generates the same error if you try and submit a non-existent template unfortunately. I knocked up a flask app that proxies the requests as they arrive on The Hive server and strips out the template line (I'll probably adjust it at some point to actually use our templates) and it works fine for generating cases now, which makes my life easier :)
